### PR TITLE
Handling window frame of rank-related functions in snowflake

### LINKED ISSUE #782: [BUG]: Snowflake SQL translations provide incorrect output for the ‘first_value’ analytic window function
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Category of Bug / Issue

TranspileParserError

### Current Behavior

On Snowflake, the default window frame specification for rank-related functions such as `first_value`, `last_value`, `nth_value` is `ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING` ([Snowflake Docs Reference](https://docs.snowflake.com/en/sql-reference/functions-analytic#usage-notes-for-window-frames))

However, on Databricks the `first_value` function is classified as an `Analytic Window function` and has a default window frame specification of `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW` when an order by clause is present. ([Stackoverflow reference](https://stackoverflow.com/questions/47130030/whats-the-default-window-frame-for-window-functions))

### Expected Behavior

The correct translation should include explicitly adding the window frame specification `ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING` to the `PARTITION BY` clause to match the default Snowflake behavior if a Window frame specification is missing in the original query.  

### Steps To Reproduce

Original Snowflake Query: 
`SELECT    tabb.col_a,
          tabb.col_b,
          First_value(
          CASE
                    WHEN tabb.col_c IN (‘xyz’,
                                        ‘abc’) THEN tabb.col_d
          END) ignore nulls OVER (partition BY tabb.col_e ORDER BY tabb.col_f) AS derived_col_a
FROM      schema_a.table_a taba
LEFT JOIN schema_b.table_b AS tabb
ON        taba.col_e = tabb.col_e`

Translated Query:
`SELECT
  tabb.col_a,
  tabb.col_b,
  FIRST_VALUE(CASE WHEN tabb.col_c IN (‘xyz’, ‘abc’) THEN tabb.col_d END) IGNORE NULLS OVER (PARTITION BY tabb.col_e ORDER BY tabb.col_f NULLS LAST) AS derived_col_a
FROM schema_a.table_a AS taba
LEFT JOIN schema_b.table_b AS tabb
  ON taba.col_e = tabb.col_e`

Expected Query to match Snowflake SQL output:
`SELECT
  tabb.col_a,
  tabb.col_b,
  FIRST_VALUE(CASE WHEN tabb.col_c IN (‘xyz’, ‘abc’) THEN tabb.col_d END) IGNORE NULLS OVER (PARTITION BY tabb.col_e ORDER BY tabb.col_f NULLS LAST ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS derived_col_a
FROM schema_a.table_a AS taba
LEFT JOIN schema_b.table_b AS tabb
  ON taba.col_e = tabb.col_e`

### Relevant log output or Exception details

_No response_

### Sample Query

```shell
SELECT    tabb.col_a,
          tabb.col_b,
          First_value(
          CASE
                    WHEN tabb.col_c IN (‘xyz’,
                                        ‘abc’) THEN tabb.col_d
          END) ignore nulls OVER (partition BY tabb.col_e ORDER BY tabb.col_f) AS derived_col_a
FROM      schema_a.table_a taba
LEFT JOIN schema_b.table_b AS tabb
ON        taba.col_e = tabb.col_e
```

### Operating System

macOS

### Version

v0.1.5

### PR DESCRIPTION
**Pull request by ganeshdogiparthi-db**
closes #782

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
still failing tests:

lag_1.sql:
```
LAG: illegal index: 1, expr: CallFunction LAG
+- Id col1, false
```

last_value_2.sql
```
NullPointerException: null
```

lead_1.sql
```
LEAD: illegal index: 1, expr: CallFunction LEAD
+- Id col1, false
```

nth_value_3.sql
```
NullPointerException: null
```

File: tests/resources/functional/snowflake/functions/window/last_value_2.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (nfx) [inline tests/resources/functional/snowflake/functions/window/last_value_2.sql:1]: remove leading blank line. looks messy

File: tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (nfx) [inline tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql:1]: remove leading blank lines from all example files. looks unprofessional.

File: tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+-- snowflake sql:
+SELECT dense_rank()  OVER (PARTITION BY col1 ORDER BY col2 DESC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS dense_rank_col1 FROM tabl;
+
+-- databricks sql:
+SELECT DENSE_RANK() OVER (PARTITION BY col1 ORDER BY col2 DESC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS dense_rank_col1 FROM tabl;
```
Comment (nfx) [inline tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql:6]: pretty-print queries/translated examples. use `pretty=True` when generating them

Comment (ganeshdogiparthi-db) [reply tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql]: Done

File: tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql (hunk 2 above)
Comment (ganeshdogiparthi-db) [reply tests/resources/functional/snowflake/aggregates/test_dense_rank_2.sql]: Done

File: tests/resources/functional/snowflake/functions/window/last_value_2.sql (hunk 1 above)
Comment (ganeshdogiparthi-db) [reply tests/resources/functional/snowflake/functions/window/last_value_2.sql]: Done

File: tests/resources/functional/snowflake/functions/window/nth_value_3.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+
+-- snowflake sql:
+SELECT tabb.col_a, tabb.col_b, nth_value(CASE WHEN tabb.col_c IN (‘xyz’, ‘abc’) THEN tabb.col_d END, 42) ignore nulls OVER (partition BY tabb.col_e ORDER BY tabb.col_f DESC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS derived_col_a FROM schema_a.table_a taba LEFT JOIN schema_b.table_b AS tabb ON taba.col_e = tabb.col_e;
+
+-- databricks sql:
+SELECT tabb.col_a, tabb.col_b, NTH_VALUE(CASE WHEN tabb.col_c IN (‘xyz’, ‘abc’) THEN tabb.col_d END, 42) IGNORE NULLS OVER (PARTITION BY tabb.col_e ORDER BY tabb.col_f DESC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS derived_col_a FROM schema_a.table_a AS taba LEFT JOIN schema_b.table_b AS tabb ON taba.col_e = tabb.col_e;
```
[the remark this answers is not in the document]
Comment (ganeshdogiparthi-db) [reply tests/resources/functional/snowflake/functions/window/nth_value_3.sql]: Done

File: core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/functions.scala [hunk 5]
```diff
@@ -144,10 +144,10 @@ class CallMapper extends IRHelpers {
     case CallFunction("JSON_OBJECT_KEYS", args) => JsonObjectKeys(args.head)
     case CallFunction("JSON_TUPLE", args) => JsonTuple(args)
     case CallFunction("KURTOSIS", args) => Kurtosis(args.head)
-    case CallFunction("LAG", args) => Lag(args.head, args(1), args(2))
+    // case CallFunction("LAG", args) => Lag(args.head, args(1), args(2))
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/functions.scala:147]:
```suggestion
```

File: core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/functions.scala [hunk 6]
```diff
@@ -144,10 +144,10 @@ class CallMapper extends IRHelpers {
...
     case CallFunction("LAST", args) => Last(args.head, args(1))
     case CallFunction("LAST_DAY", args) => LastDay(args.head)
-    case CallFunction("LEAD", args) => Lead(args.head, args(1), args(2))
+    // case CallFunction("LEAD", args) => Lead(args.head, args(1), args(2))
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/intermediate/functions.scala:150]:
```suggestion
```
remove commented code

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**